KAFKA-4441 Monitoring incorrect during topic creation and deletion

edoardocomar opened the pull request:
OfflinePartitionsCount PreferredReplicaImbalanceCount metrics check for
topic being deleted

Added integration test which polls the metrics while topics are being
created and deleted

Developed with @mimaison

ijuma commented:
cc @junrao

edoardocomar commented:
The integration/unit test we added reproduced the spurious metric value during deletion for 
`PreferredReplicaImbalanceCount`  and `OfflinePartitionsCount` but could not produce an unexpected value for `UnderReplicatedPartitions`

mimaison commented:
Actually we managed to get unexpected values for UnderReplicatedPartitions while creating and deleting topics. 

We can't fix it like the other 2 metrics because ReplicaManager, which emits this metric, does not have access to TopicDeletionManager (held in KafkaController). 

Currently in KafkaServer, ReplicaManager is created before KafkaController so we can't pass it in a reference to the TopicDeletionManager. We're actually unsure if changing the order will have side effects ...

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,149 @@
+        Metrics.defaultRegistry.allMetrics.asScala
+        .filterKeys(k => k.getName.endsWith(m))
+        .headOption match {
+           case Some(e) => Metrics.defaultRegistry.removeMetric(e._1)
```
Hmm, not sure what this code does. Could you add a comment?

edoardocomar replied:
The snippet removes old metrics from the MetricsRegistry. I'll add a comment.

It's an attempt to mitigate the problems described by https://issues.apache.org/jira/browse/KAFKA-4605
Essentially, each time the harness is run and new server/controller objects are created, the actual metric objects in the Default Metrics Registry are not re-created; old metrics are kept even though they are attached to 'dead' servers/controllers.

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,149 @@
+    }
+    
+    super.setUp
+    log.setLevel(Level.INFO)
```
Is this needed? We can set the log4j level to info by changing the log4j property file for debugging purpose.

edoardocomar replied:
It sets the loglevel only on the logger used by the test, it's not strictly needed, we can remove it though we found it useful during the debugging phase.

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,149 @@
+    assert(gaugeValue == 0)
+    
+    // Thread checking the metric continuously
+    var running = true
```
Does this and gaugeValue need to be volatile?

edoardocomar replied:
We'll add it, it's good coding, though on both Linux and Mac we had the test working anyway.

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,149 @@
+          }
+          if (!running) break
+        }
+        Thread.sleep(1000)
```
Could we reduce the amount of time to sleep?

junrao reviewed (commented):
@edoardocomar : Thanks for the patch. Left a few comments.

edoardocomar commented:
@junrao Thanks for the comments - the new PR addresses them and also has the test and a possible fix for `UnderReplicatedPartition`

please could you take another look? thanks

edoardocomar commented:
please retest

edoardocomar committed:
KAFKA-4441 Monitoring incorrect during topic creation and deletion

OfflinePartitionsCount and PreferredReplicaImbalanceCount metrics check
for topic being deleted.
Fix to Replica.updateLogResult() suggested by @lindong28

Added integration test which polls the metrics while topics are being
created and deleted

developed with @mimaison

edoardocomar commented:
@junrao could you please take another look? This is a fix we'd love to have in 10.2 thx!

edoardocomar commented:
please retest this

edoardocomar commented:
Jenkins retest this please

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -388,7 +388,7 @@ class Partition(val topic: String,
       leaderReplicaIfLocal match {
         case Some(leaderReplica) =>
-          val outOfSyncReplicas = getOutOfSyncReplicas(leaderReplica, replicaMaxLagTimeMs)
+          val outOfSyncReplicas = getOutOfSyncReplicas(leaderReplica, replicaMaxLagTimeMs).filter(r => !r.ignoreIsrCheck)
```
Hmm, could you explain why we need to do the one-time ignoreIsrCheck? In makeLeader(), we already reset the lastCaughtUpTime to the current time for in-sync replicas.

edoardocomar replied:
What our investigation found, is that, _with a newly created topic_, the isr-expiration thread may call `maybeShrinkIsr` before the lastCaughtUpTime gets updated, when its value is still -1, and therefore shrinks the ISR temporarily - and increase the metric UnderReplicatedPartitionCount, which is what the KAFKA-4441 bug report was about.

This is demonstrated by our unit test, if you remove 
`filter(r => !r.ignoreIsrCheck)`
the test will fail. QED.

The `ignoreIsrCheck` allows to ignore this check to be ignored on a new Replica for one period of the ISR scheduled thread, i.e. `period = config.replicaLagTimeMaxMs / 2`
We think that is acceptable.

edoardocomar replied:
Updated the comment in `Replica.ignoreIsrCheck()`

junrao replied:
Good point. I am wondering if a simpler fix is to pass in lastCaughtUpTimeMs when constructing the Replica object and we can initialize it to current time.

BTW, do you see this issue on 0.10.1? In 0.10.1, lastCaughtUpTimeMsUnderlying actually initializes to current time in Replica.

edoardocomar replied:
Hi @junrao we saw the metrics issue both on the current trunk (where we wrote the test that demonstrated each of the three issue) and on our 10.0.1 system where we occasionally noticed those metrics going off when topic creation/deletion was going on.
We'll take another look at the specific UnderReplicatedPartitionCount.
cc @mimaison

mimaison replied:
@junrao 
Without the fix, running the test against current trunk fails. 
The UnderReplicatedPartitionCount value briefly increases when creating topics.

I can see the code has changed since 0.10.1. In 0.10.1, as you mentioned, lastCaughtUpTime is initialized to current time whereas in trunk it's initialized to 0. See https://github.com/apache/kafka/commit/fc88bac66686ef963b5aec53a3eba717336e4ba7#diff-66ab822b4a165d4717324669c01ba2fb

I ran the test against 0.10.1 and it's not hitting the UnderReplicatedPartitionCount issue so it looks like https://github.com/apache/kafka/commit/fc88bac66686ef963b5aec53a3eba717336e4ba7 reintroduced the issue !

junrao replied:
@mimaison : Thanks for the update. For trunk, do you think you can improve the patch by initializing lastCaughtUpTimeMsUnderlying to current time instead? Then, we should be able to get rid of ignoreIsrCheck().

edoardocomar replied:
Hi @junrao
in trunk there is no longer a `lastCaughtUpTimeMsUnderlying`

if we try to initialize in `Replica`
```
  @volatile private[this] var _lastCaughtUpTimeMs = time.milliseconds
```
our `MetricsDuringTopicCreationDeletionTest` proves that the ISR still may shrink temporarily - if the isr-thread runs just after `Replica.updateLogReadResult` has been invoked with `LogReadResult.UnknownLogReadResult`

junrao replied:
@edoardocomar : Got it. Could you try something like the following (no need to initialize _lastCaughtUpTimeMs in Replica)?

diff --git a/core/src/main/scala/kafka/cluster/Partition.scala b/core/src/main/scala/kafka/cluster/Partition.scala
index 46815e1..b80fb5c 100755
--- a/core/src/main/scala/kafka/cluster/Partition.scala
+++ b/core/src/main/scala/kafka/cluster/Partition.scala
@@ -190,7 +190,8 @@ class Partition(val topic: String,
         // construct the high watermark metadata for the new leader replica
         leaderReplica.convertHWToLocalOffsetMetadata()
         // reset log end offset for remote replicas
-        assignedReplicas.filter(_.brokerId != localBrokerId).foreach(_.updateLogReadResult(LogReadResult.UnknownLogReadResult))
+        val unknownLogReadResultWithCurTime = LogReadResult.UnknownLogReadResult.copy(fetchTimeMs = curTimeMs)
+        assignedReplicas.filter(_.brokerId != localBrokerId).foreach(_.updateLogReadResult(unknownLogReadResultWithCurTime))
       }
       (maybeIncrementLeaderHW(leaderReplica), isNewLeader)
     }

mimaison replied:
@junrao Thanks for the suggestion.
I gave it a go and it makes our test pass without the need for ignoreIsrCheck() but it breaks a number of tests in UncleanLeaderElectionTest and ReplicationQuotasTest. I'm not too familiar with these tests and couldn't tell immediately what's going on. I'll have a closer look see if I can understand what breaks them

edoardocomar replied:
@mimaison  I had those same test failures when I tried initialising to `lastCaughtUpTimeMs` to currentTime in both the variable declaration and in the  `Replica.updateLogReadResult(..)` method when invoked with the `UnknownLogReadResult`, in what was essentially @junrao suggestion

junrao replied:
@lindong28 : Dong, since you added this logic, do you think the above suggested approach works? If so, do you have any insight on why it may cause some existing tests to break?

edoardocomar replied:
Hi @lindong28 @junrao how could we make progress on this JIRA and have the fix merged in trunk ?
thanks

lindong28 replied:
@junrao @edoardocomar oops.. Sorry for delayed response. I didn't notice that I am pinged in this thread until now. Let me take a look at this patch.

lindong28 replied:
@junrao @edoardocomar Sorry for the delayed response and this bug. I think I know the problem now.

I don't think there is any problem with initializing `_lastCaughtUpTimeMs` to 0L in Replica. When a topic is created, controller will send `LeaderAndIsrRequest` with isr=all_replicas. When leader receives this `LeaderAndIsrRequest`, it will call `Partition.makeLeader()`, which does the following:

```
(assignedReplicas - leaderReplica).foreach { replica =>
  val lastCaughtUpTimeMs = if (inSyncReplicas.contains(replica)) curTimeMs else 0L
  replica.resetLastCaughtUpTime(curLeaderLogEndOffset, curTimeMs, lastCaughtUpTimeMs)
}
```

Because all replicas are in isr set, their last caughtup time will be initialized to current time. Up to now we are doing the right things. Because controller determines isr set when a broker is newly elected leader, and a replica's `_lastCaughtUpTimeMs` should be initialized to current time if it is in sync. Otherwise we set `_lastCaughtUpTimeMs` to 0.

The problem is that we does `assignedReplicas.filter(_.brokerId != localBrokerId).foreach(_.updateLogReadResult(LogReadResult.UnknownLogReadResult))`, which sets `lastCaughtUpTimeMs` of all replicas to -1. I missed this when I implemented the patch earlier.

I think the right fix is to do the following in `updateLogResult(...)`. I have verified that the test created in this patch can succeed after this fix.

```
    if (logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.leaderLogEndOffset)
      _lastCaughtUpTimeMs = math.max(_lastCaughtUpTimeMs, logReadResult.fetchTimeMs)
    else if (logReadResult.info.fetchOffsetMetadata.messageOffset >= lastFetchLeaderLogEndOffset)
      _lastCaughtUpTimeMs = math.max(_lastCaughtUpTimeMs, lastFetchTimeMs)
```

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,163 @@
+import kafka.utils._
+import scala.collection.JavaConverters.mapAsScalaMapConverter
+
+import org.apache.log4j.{Level, Logger}
```
unused imports. "After" is also not used.

edoardocomar replied:
thanks, relics of old code; sadly, Eclipse lacks a Scala warning

junrao commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -0,0 +1,163 @@
+
+    // if the thread checking the gauge is still run, stop it
+    running = false;
+    thread.interrupt
```
Do we need to interrupt? There is no blocking operation in the thread. So it seems just checking running is enough.

edoardocomar replied:
thanks, `interrupt` no longer needed

junrao reviewed (commented):
@edoardocomar : Sorry for the delay. Left a few more comments.

edoardocomar commented:
Many thanks @lindong28 we verified your suggestion, great.

Code updated and rebased.

ijuma commented on core/src/main/scala/kafka/cluster/Replica.scala:
```diff
@@ -130,6 +130,7 @@ class Replica(val brokerId: Int,
     replicaString.append("; Topic: " + partition.topic)
     replicaString.append("; Partition: " + partition.partitionId)
     replicaString.append("; isLocal: " + isLocal)
+    replicaString.append("; lastCUTms: " + _lastCaughtUpTimeMs)
```
Typo.

edoardocomar replied:
Hi @ijuma do you prefer the `toString` method to list the `_lastCaughtUpTimeMs` field without abbreviation?

ijuma replied:
Yes, `lastCUTms` is very obscure.

edoardocomar replied:
can't disagree :-)

edoardocomar committed:
KAFKA-4441 Monitoring incorrect during topic creation and deletion

OfflinePartitionsCount and PreferredReplicaImbalanceCount metrics check
for topic being deleted.
Fix to Replica.updateLogResult() suggested by @lindong28

Added integration test which polls the metrics while topics are
being created and deleted

developed with @mimaison

lindong28 commented:
Thanks for the patch. LGTM.

junrao commented:
@edoardocomar : Thanks for the patch. LGTM too. @lindong28 : Thanks for your help!.
